In [2]:
import numpy as np    #importar numpy
import pandas as pd   #importar pandas

In [3]:
wdata=pd.read_csv('w_data.csv',sep=';')   #cargar el archivo csv w_data
wdata.head(10)                              #mostrar las primeras lineas de w_data

Workclass    fnlwgt  Capital Gain  Capital Loss  Hours per week  \
0          State-gov   77516.0        2174.0           0.0            40.0   
1   Self-emp-not-inc   83311.0           0.0           0.0            13.0   
2            Private  215646.0           0.0           0.0            40.0   
3            Private  234721.0           0.0           0.0            40.0   
4            Private  338409.0           0.0           0.0            40.0   
5            Private  284582.0           0.0           0.0            40.0   
6            Private  160187.0           0.0           0.0            16.0   
7   Self-emp-not-inc  209642.0           0.0           0.0            45.0   
8            Private   45781.0       14084.0           0.0            50.0   
9            Private  159449.0        5178.0           0.0            40.0   

       ID  Target  
0   49084   <=50K  
1   68553   <=50K  
2  143654   <=50K  
3   41510   <=50K  
4  195025   <=50K  
5  231557   <=50K  
6  393526   <=50K  
7  488425    >50K  
8    7496    >50K  
9  576313    >50K

In [4]:
wdata.info()                   #ver info del dataset, donde hay 48843 datos con ID, y para el resto de atributos hay 48842

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48843 entries, 0 to 48842
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Workclass       48842 non-null  object 
 1   fnlwgt          48842 non-null  float64
 2   Capital Gain    48842 non-null  float64
 3   Capital Loss    48842 non-null  float64
 4   Hours per week  48842 non-null  float64
 5   ID              48843 non-null  int64  
 6   Target          48842 non-null  object 
dtypes: float64(4), int64(1), object(2)
memory usage: 2.6+ MB


In [22]:
pdata=pd.read_csv('p_data.csv',sep=';')    #cargar el archivo csv w_data
pdata.head(5)                               #mostrar las primeras lineas de p_data

Age      Sex      ID    Race         Country   Education  Education-Num  \
0  39.0     Male   49084   White   United-States   Bachelors           13.0   
1  50.0     Male   68553   White   United-States   Bachelors           13.0   
2  38.0     Male  143654   White   United-States     HS-grad            9.0   
3  53.0     Male   41510   Black   United-States        11th            7.0   
4  28.0   Female  195025   Black            Cuba   Bachelors           13.0   

        Marital Status    Relationship  
0        Never-married   Not-in-family  
1   Married-civ-spouse         Husband  
2             Divorced   Not-in-family  
3   Married-civ-spouse         Husband  
4   Married-civ-spouse            Wife

In [37]:
pdata.info()                   #ver info del dataset, donde hay 48842 datos con ID, y para el resto de atributos hay 48841

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48842 entries, 0 to 48841
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Age             48841 non-null  float64
 1   Sex             48841 non-null  object 
 2   ID              48842 non-null  int64  
 3   Race            48841 non-null  object 
 4   Country         48841 non-null  object 
 5   Education       48841 non-null  object 
 6   Education-Num   48841 non-null  float64
 7   Marital Status  48841 non-null  object 
 8   Relationship    48841 non-null  object 
dtypes: float64(2), int64(1), object(6)
memory usage: 3.4+ MB


In [24]:
pdata['ID'].value_counts()

511184898     1
1081672469    1
362315312     1
268622123     1
586504157     1
             ..
205790697     1
541328874     1
628362731     1
1844323822    1
324339711     1
Name: ID, Length: 48842, dtype: int64

In [25]:
result=pd.merge(wdata, pdata, on='ID', how='outer',indicator='union')      #fusion de ambos datasets por medio del atributo en comun 'ID', llamandolo 'result'. Se incluye una columna con la union que indica que si esta en ambos datasets o no
result.head()                                                              #mostrar las primeras lineas de result o informacion fusionada de w_data y p_data

Workclass    fnlwgt  Capital Gain  Capital Loss  Hours per week  \
0          State-gov   77516.0        2174.0           0.0            40.0   
1   Self-emp-not-inc   83311.0           0.0           0.0            13.0   
2            Private  215646.0           0.0           0.0            40.0   
3            Private  234721.0           0.0           0.0            40.0   
4            Private  338409.0           0.0           0.0            40.0   

       ID  Target   Age      Sex    Race         Country   Education  \
0   49084   <=50K  39.0     Male   White   United-States   Bachelors   
1   68553   <=50K  50.0     Male   White   United-States   Bachelors   
2  143654   <=50K  38.0     Male   White   United-States     HS-grad   
3   41510   <=50K  53.0     Male   Black   United-States        11th   
4  195025   <=50K  28.0   Female   Black            Cuba   Bachelors   

   Education-Num       Marital Status    Relationship union  
0           13.0        Never-married   Not-in-family  both  
1           13.0   Married-civ-spouse         Husband  both  
2            9.0             Divorced   Not-in-family  both  
3            7.0   Married-civ-spouse         Husband  both  
4           13.0   Married-civ-spouse            Wife  both

In [26]:
result.info()               #la columna ID en wdata trae 48843 valores, y en pdata trae 48842, por lo tanto hay 1 valor diferente

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48843 entries, 0 to 48842
Data columns (total 16 columns):
 #   Column          Non-Null Count  Dtype   
---  ------          --------------  -----   
 0   Workclass       48842 non-null  object  
 1   fnlwgt          48842 non-null  float64 
 2   Capital Gain    48842 non-null  float64 
 3   Capital Loss    48842 non-null  float64 
 4   Hours per week  48842 non-null  float64 
 5   ID              48843 non-null  int64   
 6   Target          48842 non-null  object  
 7   Age             48841 non-null  float64 
 8   Sex             48841 non-null  object  
 9   Race            48841 non-null  object  
 10  Country         48841 non-null  object  
 11  Education       48841 non-null  object  
 12  Education-Num   48841 non-null  float64 
 13  Marital Status  48841 non-null  object  
 14  Relationship    48841 non-null  object  
 15  union           48843 non-null  category
dtypes: category(1), float64(6), int64(1), object(8)
memory usa

In [27]:
result['union'].value_counts()    #cuenta los valores que coinciden en ambos datasets, y el que esta solo en el dataset de la izquierda

both          48842
left_only         1
right_only        0
Name: union, dtype: int64

##### En total hay 48843 valores de ID, y coinciden en ambas tablas 48842; mientras que 1 de los datos esta en wdata, pero no en pdata

In [28]:
result[result['union']!='both']['ID']    #al filtrar por el nuevo atributo union, el dato que sea diferente de both es el ID 1344995623

48842    1344995623
Name: ID, dtype: int64

##### El valor que no es comun es el 1344995623

In [29]:
result['Target'].unique()    #usar unique para conocer los posibles valores en el atributo Target

array([' <=50K', ' >50K', nan, ' <=50K.', ' >50K.'], dtype=object)

In [38]:
result

Workclass    fnlwgt  Capital Gain  Capital Loss  \
0              State-gov   77516.0        2174.0           0.0   
1       Self-emp-not-inc   83311.0           0.0           0.0   
2                Private  215646.0           0.0           0.0   
3                Private  234721.0           0.0           0.0   
4                Private  338409.0           0.0           0.0   
...                  ...       ...           ...           ...   
48838            Private  215419.0           0.0           0.0   
48839                  ?  321403.0           0.0           0.0   
48840            Private  374983.0           0.0           0.0   
48841            Private   83891.0        5455.0           0.0   
48842       Self-emp-inc  182148.0           0.0           0.0   

       Hours per week          ID  Target   Age      Sex                 Race  \
0                40.0       49084   <=50K  39.0     Male                White   
1                13.0       68553   <=50K  50.0     Male                White   
2                40.0      143654   <=50K  38.0     Male                White   
3                40.0       41510   <=50K  53.0     Male                Black   
4                40.0      195025   <=50K  28.0   Female                Black   
...               ...         ...     ...   ...      ...                  ...   
48838            36.0   735617153   <=50K  39.0   Female                White   
48839            40.0  1131807446   <=50K  64.0     Male                Black   
48840            50.0   781422087   <=50K  38.0     Male                White   
48841            40.0   161771296   <=50K  44.0     Male   Asian-Pac-Islander   
48842            60.0  1344995623    >50K   NaN      NaN                  NaN   

              Country   Education  Education-Num       Marital Status  \
0       United-States   Bachelors           13.0        Never-married   
1       United-States   Bachelors           13.0   Married-civ-spouse   
2       United-States     HS-grad            9.0             Divorced   
3       United-States        11th            7.0   Married-civ-spouse   
4                Cuba   Bachelors           13.0   Married-civ-spouse   
...               ...         ...            ...                  ...   
48838   United-States   Bachelors           13.0             Divorced   
48839   United-States     HS-grad            9.0              Widowed   
48840   United-States   Bachelors           13.0   Married-civ-spouse   
48841   United-States   Bachelors           13.0             Divorced   
48842             NaN         NaN            NaN                  NaN   

          Relationship      union  
0        Not-in-family       both  
1              Husband       both  
2        Not-in-family       both  
3              Husband       both  
4                 Wife       both  
...                ...        ...  
48838    Not-in-family       both  
48839   Other-relative       both  
48840          Husband       both  
48841        Own-child       both  
48842              NaN  left_only  

[48843 rows x 16 columns]

In [30]:
result['Target']=result['Target'].replace([' <=50K.',' >50K.'],[' <=50K',' >50K'])

##### A continuacion se muestra el dataset corregido

In [31]:
result

Workclass    fnlwgt  Capital Gain  Capital Loss  \
0              State-gov   77516.0        2174.0           0.0   
1       Self-emp-not-inc   83311.0           0.0           0.0   
2                Private  215646.0           0.0           0.0   
3                Private  234721.0           0.0           0.0   
4                Private  338409.0           0.0           0.0   
...                  ...       ...           ...           ...   
48838            Private  215419.0           0.0           0.0   
48839                  ?  321403.0           0.0           0.0   
48840            Private  374983.0           0.0           0.0   
48841            Private   83891.0        5455.0           0.0   
48842       Self-emp-inc  182148.0           0.0           0.0   

       Hours per week          ID  Target   Age      Sex                 Race  \
0                40.0       49084   <=50K  39.0     Male                White   
1                13.0       68553   <=50K  50.0     Male                White   
2                40.0      143654   <=50K  38.0     Male                White   
3                40.0       41510   <=50K  53.0     Male                Black   
4                40.0      195025   <=50K  28.0   Female                Black   
...               ...         ...     ...   ...      ...                  ...   
48838            36.0   735617153   <=50K  39.0   Female                White   
48839            40.0  1131807446   <=50K  64.0     Male                Black   
48840            50.0   781422087   <=50K  38.0     Male                White   
48841            40.0   161771296   <=50K  44.0     Male   Asian-Pac-Islander   
48842            60.0  1344995623    >50K   NaN      NaN                  NaN   

              Country   Education  Education-Num       Marital Status  \
0       United-States   Bachelors           13.0        Never-married   
1       United-States   Bachelors           13.0   Married-civ-spouse   
2       United-States     HS-grad            9.0             Divorced   
3       United-States        11th            7.0   Married-civ-spouse   
4                Cuba   Bachelors           13.0   Married-civ-spouse   
...               ...         ...            ...                  ...   
48838   United-States   Bachelors           13.0             Divorced   
48839   United-States     HS-grad            9.0              Widowed   
48840   United-States   Bachelors           13.0   Married-civ-spouse   
48841   United-States   Bachelors           13.0             Divorced   
48842             NaN         NaN            NaN                  NaN   

          Relationship      union  
0        Not-in-family       both  
1              Husband       both  
2        Not-in-family       both  
3              Husband       both  
4                 Wife       both  
...                ...        ...  
48838    Not-in-family       both  
48839   Other-relative       both  
48840          Husband       both  
48841        Own-child       both  
48842              NaN  left_only  

[48843 rows x 16 columns]

In [32]:
result['Target'].value_counts()      #cuenta la cantidad de IDs que son <=50 y los que son >50

 <=50K    37155
 >50K     11687
Name: Target, dtype: int64

In [33]:
result['Target'].value_counts()/len(result)*100      #al dividir por el total y multiplicar por 100, tenemos los % de cada grupo

 <=50K    76.070266
 >50K     23.927687
Name: Target, dtype: float64

#### El resultado es que el 76.07% ganan un valor menor o igual a 50K, y el 23.93% restante gana mas de 50K

In [34]:
PT_Target_Sex=(result.pivot_table(values=['ID'],index='Target',columns=['Sex'],aggfunc='count'))/len(result)*100   #arma una pivot table con el target en las filas, el sexo en las columnas y cuenta los disntintos valores que tienen.  Luego se calcula el porcentaje al dividir cada valor por la longitud total del dataset
PT_Target_Sex                                                                                          #muestra PT_Target_Sex (PT:Pivot Table)

ID           
Sex        Female       Male
Target                      
 <=50K  29.529308  46.540958
 >50K    3.621809  20.303831

##### en el rango de ganancias menores o iguales a 50K, se encuentra el 29.52% de las mujeres y el 46.54% de los hombres
      en el rango de ganancias mayores a 50K, se encuentra el 3.62% de las mujeres y el 20.30% de los hombres

In [35]:
PT_Target_Race=(result.pivot_table(values=['ID'],index='Target',columns=['Race'],aggfunc='count'))/len(result)*100   #arma una pivot table con el target en las filas, la raza en las columnas y cuenta los disntintos valores que tienen.  Luego se calcula el porcentaje al dividir cada valor por la longitud total del dataset
PT_Target_Race                       #muestra PT_Target_Race (PT:Pivot Table)

ID                                                   
Race    Amer-Indian-Eskimo  Asian-Pac-Islander     Black     Other      White
Target                                                                       
 <=50K            0.849661            2.272588  8.433143  0.728866  63.786008
 >50K             0.112606            0.837377  1.158815  0.102369  21.714473

##### los mayores porcentajes se encuentran en personas de raza blanca (white), donde el 63.78% ganan 50K o menos, y el 21.71% ganan mas de 50K

In [36]:
PT_Education_Target=(result.pivot_table(values=['ID'],index='Education',columns=['Target'],aggfunc='count'))/len(result)*100   #arma una pivot table con la educacion en las filas, el target en las columnas y cuenta los disntintos valores que tienen.  Luego se calcula el porcentaje al dividir cada valor por la longitud total del dataset
PT_Education_Target

ID          
Target             <=50K      >50K
Education                         
 10th           2.665684  0.178122
 11th           3.521487  0.188359
 12th           1.246852  0.098274
 1st-4th        0.489323  0.016379
 5th-6th        0.986835  0.055279
 7th-8th        1.828307  0.126937
 9th            1.463874  0.083942
 Assoc-acdm     2.432283  0.845566
 Assoc-voc      3.150912  1.068730
 Bachelors      9.647237  6.780910
 Doctorate      0.333722  0.882419
 HS-grad       27.191204  5.124583
 Masters        2.452757  2.987122
 Preschool      0.167885  0.002047
 Prof-school    0.444281  1.263231
 Some-college  18.047622  4.223737

##### en el rango de ganancias menores o iguales a 50K, la mayor proporcion la tienen los HS-grad 27.19% y Some-college 18.04%
      en el rango de ganancias mayores a 50K, la mayor proporcion la tienen los bachelors con 6.78%